python/ambassador/envoy/v3/v3listener.py

```diff
@@ -1008,25 +1017,23 @@ def __str__(self) -> str:
     @classmethod
     def generate(cls, config: 'V3Config') -> None:
         config.listeners = []
-        logger = config.ir.logger
 
         for key in config.ir.listeners.keys():
             irlistener = config.ir.listeners[key]
             v3listener = V3Listener(config, irlistener)
             v3listener.finalize()
 
-            config.ir.logger.info(f"V3Listener: ==== GENERATED {v3listener}")
-
-            if v3listener._log_debug:
-                for k in sorted(v3listener._chains.keys()):
-                    chain = v3listener._chains[k]
-                    config.ir.logger.debug("    %s", chain)
-
-                    for r in chain.routes:
-                        config.ir.logger.debug("      %s", v3prettyroute(r))
+            config.ir.logger.info(f"V3Listener {v3listener}: generated ===========================")
+            if config.ir.logger.isEnabledFor(logging.DEBUG):
+                if v3listener._log_debug:
```

Comment:
There's a pre-existing comment somewhere in the file about this, but to say it again:

Python will process the `f"string"` formatting or `"fmt" % args` formatting before passing it to `.debug(…)`; even if `.debug(…)` is just going to discard it.  Someone did some benchmarks and figured out that we were spending a _lot_ of CPU formatting debug log strings even when the logger was just going to discard them.

So, check the log-level before formatting the log string.

Reply:
Ok, the reducing CPU cycles for formatting strings that are going to be discarded I understand. I guess the part I'm still missing is the why is `config.ir.logger.isEnabledFor(logging.DEBUG)` not sufficient? Why does `v3listener` have its own `._log_debug` that needs to be checked?

Reply:
I assume because it's shorter to type out. Perhaps I should have made that an `@property` like I did other things.

Reply:
just to add even more the needless complexity of python logging, f-strings squash things to a single string that gets added to the Log record so usually prefer to use old-school `%s` because the logging library is optimized to defer formatting until it needs to print it out thought can change the formatter on your logger so it's basically the equivalent of the if statement above though I haven't actually benchmarked this. 

It's also relevant when you use error log aggregators like Sentry for cleaner error log groupings.

Reply:
@haq204 So the lesson isn't "go back to using old-school `log("str" % args`)", it's "use old-school-style 'str' with `log("str", args)`. Good to know.

Reply:
Is there a linter that can check for formatted strings (f-strings, or `%` expressions) being passed to the logger, instead of passing the args as args?

Reply:
@LukeShu I think pytlint has a check for that.

Reply:
And, FWIW, I wouldn't say that's complicated, it's a totally expected optimization (we do it in dlog!), just one that apparently none of us realized Python makes available to us.

Reply:
yea I meant more that if you look at the internals of python logging its much more complicated than what a lot of people think it is

Reply:
https://github.com/emissary-ingress/emissary/issues/4520

Reply:
Several of "we" did know about that – in many places you will see "%s" formatting in logging instead of f-strings for this reason. Here's the gotcha though:

Good: `log.Debug("oh no, the horrible %s %s %s!", "grue", "ate", "Jane")`
Bad: `log.Debug("oh no, the horrible %s %s %s!" % ("grue", "ate", "Jane"))`

So the linter needs to check for the difference between those two.